core/src/main/java/org/apache/calcite/sql/validate/AggregatingSelectScope.java

```diff
@@ -110,8 +111,11 @@ private Resolved resolve() {
 
     final Set<ImmutableBitSet> flatGroupSets =
         Sets.newTreeSet(ImmutableBitSet.COMPARATOR);
+    final Map<ImmutableBitSet, Integer> groupSetCount = new HashMap<>();
     for (List<ImmutableBitSet> groupSet : Linq4j.product(builder.build())) {
-      flatGroupSets.add(ImmutableBitSet.union(groupSet));
+      final ImmutableBitSet set = ImmutableBitSet.union(groupSet);
+      groupSetCount.put(set, groupSetCount.getOrDefault(set, 0) + 1);
+      flatGroupSets.add(set);
```

Comment:
Is `flatGroupSets` still valid? `groupSetCount` and `flatGroupSets` seem to duplicate each other.

Reply:
I agree. We can unify HashMap and the sorted Set as TreeMap.